mod-source-record-storage-server/src/main/java/org/folio/consumers/AuthorityDomainKafkaHandler.java

```diff
@@ -0,0 +1,103 @@
+import java.util.Collections;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.folio.dao.util.RecordType;
+import org.folio.kafka.AsyncRecordHandler;
+import org.folio.rest.jooq.enums.RecordState;
+import org.folio.services.RecordService;
+import org.folio.services.util.KafkaUtil;
+import org.springframework.stereotype.Component;
+
+@Component
+public class AuthorityDomainKafkaHandler implements AsyncRecordHandler<String, String> {
+
+  private static final Logger log = LogManager.getLogger();
+
+  private static final String DOMAIN_EVENT_TYPE_HEADER = "domain-event-type";
+  private static final String DELETE_DOMAIN_EVENT_TYPE = "DELETE";
+  private static final String DELETE_EVENT_SUB_TYPE_FLD = "deleteEventSubType";
+  private static final String TENANT_FLD = "tenant";
+
+  private final RecordService recordService;
+
+  public AuthorityDomainKafkaHandler(RecordService recordService) {
+    this.recordService = recordService;
+  }
+
+  @Override
+  public Future<String> handle(KafkaConsumerRecord<String, String> consumerRecord) {
+    log.trace("handle:: Handling kafka record: '{}'", consumerRecord);
+    String authorityId = consumerRecord.key();
+    if (isUnexpectedDomainEvent(consumerRecord)) {
+      log.trace("handle:: Expected only {} domain type. Skipping authority domain kafka record [ID: '{}']",
+        DELETE_DOMAIN_EVENT_TYPE, authorityId);
+    }
+
+    var eventPayload = new JsonObject(consumerRecord.value());
+    var tenantId = eventPayload.getString(TENANT_FLD);
+    var eventSubType = EventSubType.valueOf(eventPayload.getString(DELETE_EVENT_SUB_TYPE_FLD));
+
+    logInput(authorityId, eventSubType, tenantId);
+    return (switch (eventSubType) {
+      case SOFT_DELETE -> performSoftDelete(authorityId, tenantId);
+      case HARD_DELETE -> performHardDelete(authorityId, tenantId);
+    }).onFailure(throwable -> logError(authorityId, eventSubType, tenantId));
+  }
+
+  private Future<String> performSoftDelete(String authorityId, String tenantId) {
+    logInput(authorityId, EventSubType.SOFT_DELETE, tenantId);
+    var condition = filterRecordByExternalId(authorityId);
+    return recordService.getRecords(condition, RecordType.MARC_AUTHORITY, Collections.emptyList(), 0, 1, tenantId)
+      .compose(recordCollection -> {
+        if (recordCollection.getRecords().isEmpty()) {
+          log.debug("handle:: No records found [externalId: '{}', tenantId: '{}']", authorityId, tenantId);
+          return Future.succeededFuture();
+        }
+        var matchedId = recordCollection.getRecords().get(0).getMatchedId();
+
+        return recordService.updateRecordsState(matchedId, RecordState.DELETED, RecordType.MARC_AUTHORITY, tenantId);
+      }).map(authorityId);
+  }
+
+  private Future<String> performHardDelete(String authorityId, String tenantId) {
+    return recordService.deleteRecordsByExternalId(authorityId, tenantId).map(authorityId);
```

Comment:
is authority id and external Id the same thing?

Reply:
Yes

Reply:
Is it possible for an authority to have an external id that is not the authority id?

Reply:
I don't think such case exists.